### PR TITLE
[FLINK-4478] [flip-6] Add HeartbeatManager

### PR DESCRIPTION
**Pull request by tillrohrmann**
Adds a `HeartbeatManager` abstraction which can monitor heartbeat targets. Whenever no heartbeat signal has been received for a given heartbeat timeout interval, the heartbeat manager will issue a heartbeat timeout notification.

The `HeartbeatManager` abstraction works completely independent of the RPC service. Furthermore, it reuses the same heartbeat logic for the sending and receiving side of the heartbeat.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatListener.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	 */
+	void notifyHeartbeatTimeout(ResourceID resourceID);
+
+	/**
+	 * Callback which is called whenever a heartbeat with an associated payload is received. The
+	 * carried payload is given to this method.
+	 *
+	 * @param payload Payload of the received heartbeat
+	 */
+	void reportPayload(I payload);
```
Comment (beyond1920) [inline flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatListener.java]: maybe we should add ResourceID param to identify which sender send this incoming message

Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatListener.java]: Yes, good point. Will add it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatListener.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ */
+public interface HeartbeatListener<I, O> {
+
+	/**
+	 * Callback which is called if a heartbeat for the machine identified by the given resource
+	 * ID times out.
+	 *
+	 * @param resourceID Resource ID of the machine whose heartbeat has timed out
+	 */
+	void notifyHeartbeatTimeout(ResourceID resourceID);
```
Comment (beyond1920) [inline flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatListener.java]: It seems that JobManager and ResourceManager don't have ResourceID, only TaskExecutor has ResourceID. Would it be more proper to use something else to identify the heartbeat target?

Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatListener.java]: We need something unique. If they don't have one, then we should think about either assigning them one or to use something different. Should be easy to change in the `HeartbeatManager` implementation.

**Comment by tillrohrmann** [comment]
I've rebased the PR onto the latest flip-6 branch.

**Comment by mxm** [comment]
Could you rebase this again? I think we should merge this soon.

**Comment by tillrohrmann** [comment]
I've rebased the PR on the latest Flip-6 branch.

File: flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatManagerImpl.java [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+		public boolean isCanceled() {
+			return state.get() == State.CANCELED;
+		}
+
+		@Override
+		public void run() {
+			// The heartbeat has timed out if we're in state running
+			if (state.compareAndSet(State.RUNNING, State.TIMEOUT)) {
+				heartbeatListener.notifyHeartbeatTimeout(resourceID);
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatManagerImpl.java]: Took me a while to figure out this is called by the executor after a timeout. Makes sense.

File: flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatManagerImpl.java [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+		void resetHeartbeatTimeout(long heartbeatTimeout) {
+			if (state.get() == State.RUNNING) {
+				cancelTimeout();
+
+				futureTimeout = scheduledExecutorService.schedule(this, heartbeatTimeout, TimeUnit.MILLISECONDS);
+
+				// Double check for concurrent accesses (e.g. a firing of the scheduled future)
+				if (state.get() != State.RUNNING) {
+					cancelTimeout();
+				}
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatManagerImpl.java]: This looks like an optimization. Is is really necessary? We can still change state after we have checked.

File: flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotProtocolTest.java [hunk 5]
```diff
@@ -43,12 +42,9 @@
 import org.junit.BeforeClass;
 import org.junit.Test;
 import org.mockito.Mockito;
-import org.mockito.invocation.InvocationOnMock;
-import org.mockito.stubbing.Answer;
 
 import java.util.Collections;
 import java.util.UUID;
-import java.util.concurrent.CompletableFuture;
```
Comment (mxm) [inline flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotProtocolTest.java]: You can leave them though. Not really an issue.

File: flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatManagerImpl.java (hunk 4 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatManagerImpl.java]: If the state changes after we've created the new timeout, this means that we've already fired or canceled in the meantime. That's why we have to cancel the new scheduled timeout action. Of course, this is more of a corner case.

**Comment by tillrohrmann** [comment]
Thanks for the review @mxm. Concerning your questions:

1. Double checking is necessary due to concurrent accesses. Of course one could also use locks but that way it is less blocking.
2. I'll add a test for the `unmonitoring`.

After adding the additional test, I'll merge this PR.

**Pull request closed without merging**